File: typescript/day03.ts
import { group } from 'console';
import { loadInput } from './index'

const scores = [
    '', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
]

export const partOne = async () => {
    const data = await loadInput('03')

    let totalScore = 0;

    for (const line of data.toString().split(/[\n]/)) {
        const [compartmentOne, compartmentTwo] = [line.substring(0, line.length / 2), line.substring(line.length / 2)]

        const commonItems = new Set([...compartmentOne].filter((c) => compartmentTwo.includes(c)))

        const priority = Array.from(commonItems).length ? Array.from(commonItems).map((c) => scores.indexOf(c)).reduce((p, c) => p + c) : 0

        totalScore += priority
    }

    return totalScore
}

export const partTwo = async () => {
    const data = await loadInput('03')
    const lines = data.toString().split(/[\n]/)
    
    let totalScore = 0;
    let _groups: string[][] = []

    lines.forEach((l, index) => {
        let i = index % 3 

        if (i === 0) {
            _groups[index] = [lines[index], lines[index + 1], lines[index + 2]]
        }
    })

    const groups = _groups.filter((g) => g?.length)

    for (const group of groups) {
        const commonItems = new Set([...group[0]].filter((c) => group[1].includes(c) && group[2].includes(c)))
        
        if (commonItems?.size > 0) {
            totalScore += Array.from(commonItems).length ? Array.from(commonItems).map((c) => scores.indexOf(c)).reduce((p, c) => p + c) : 0
        }
    }

    return totalScore
}

partOne().then((result) => console.log(`Part one: ${result}`))
partTwo().then((result) => console.log(`Part two: ${result}`))

File: dart/lib/day03.dart
import 'dart:async';
import 'package:aoc2022/utils.dart';
import 'package:collection/collection.dart';

// Build a list of lowercase + uppercase letters as our scoring reference
// i.e the score of a char is indexOf(char) + 1 (since it is 0-indexed)
List<String> chars =
    List.generate(26, (index) => String.fromCharCode(index + 97))
      ..addAll(List.generate(26, (index) => String.fromCharCode(index + 65)));

Future<int> partone() async {
  var lines = await loadInputLines("03");

  var duplicates = lines.map((line) {
    // Split into 2 groups
    Set<String> sectionone =
        Set.from(line.substring(0, line.length ~/ 2).split(''));
    Set<String> sectiontwo =
        Set.from(line.substring(line.length ~/ 2, line.length).split(''));

    // Find the intersection
    Set<String> unique = sectionone.intersection(sectiontwo);

    // Score entries based on their char index in 'chars'
    return unique.fold(0, (acc, current) => acc + chars.indexOf(current) + 1);
  });

  return duplicates.reduce((value, element) => value + element);
}

Future<int> parttwo() async {
  var lines = await loadInputLines("03");

  // Segment the data into groups of three then find their unique item
  var itemPriorities = lines.slices(3).map((group) {
    var items = group.map((e) => Set<String>.from(e.split('')));

    // Take the intersection of (1 and 2) and 3
    Set<String> firstTwo = items.first.intersection(items.elementAt(1));
    Set<String> unique = firstTwo.intersection(items.last);

    // Score entries based on their char index in 'chars'
    return unique.fold(0, (acc, current) => acc + chars.indexOf(current) + 1);
  });

  return itemPriorities.reduce((value, element) => value + element);
}
